### PR TITLE
[FIX] Merge: work with sparse

### LINKED ISSUE #2155: Merge should work with sparse
##### Orange version

3.5.dev

##### Expected behavior
Merge works with sparse.

##### Actual behavior
Merge fails on sparse.

##### Steps to reproduce the behavior
Connect Corpus and File(iris) to Merge.

### PR DESCRIPTION
**Pull request by jerneju**
##### Issue
Fixes #2155.

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by jerneju** [comment]
Waiting #2286 because I need `hstack` from _util.py_.

**Comment by nikicc** [comment]
#2286 is merged. Are there some changes still needed? Can we do something about radon?

**Comment by jerneju** [comment]
@nikicc : Radon does not complain anymore because this code now uses universal `hstack`.

File: Orange/widgets/data/owmergedata.py [hunk 1]
```diff
@@ -362,20 +364,34 @@ def _join_table_by_indices(self, reduced_extra, indices):
...
-            to_change[known] = lookup[indices[known]]
+            arr1[:, [sc for sc in string_cols if sc < left_width]] = ""
+            arr2[:, [sc - left_width for sc in string_cols if sc >= left_width]] = ""
+
+        for ind, to_change, lookup in (
+                (indices[0], arr1, left),
+                (indices[1], arr2, right)):
+            known = ind != -1
+            if sum(known):
+                to_change[known] = lookup[ind[known]]
```
Comment (nikicc) [inline Orange/widgets/data/owmergedata.py]: This is probably not sufficient for sparse data. For dense, we initially set all values as missing and hence it's ok to set only known values. For sparse, we initially set all values as zeros. Hence we should, along setting known values, also set unknown values as nans.

File: Orange/widgets/data/owmergedata.py [hunk 2]
```diff
@@ -362,20 +364,38 @@ def _join_table_by_indices(self, reduced_extra, indices):
     def _join_array_by_indices(left, right, indices, string_cols=None):
         """Join (horizontally) two arrays, taking pairs of rows given in indices
         """
+        def matrix(height, width, dtype, sparse=False):
+            if not sparse:
+                return np.full((height, width), np.nan, dtype=dtype)
+            else:
+                return sp.csr_matrix((height, width), dtype=dtype)
```
Comment (kernc) [inline Orange/widgets/data/owmergedata.py]: Emits a warning in the for loop below:
> SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.

File: Orange/widgets/data/owmergedata.py [hunk 3]
```diff
@@ -362,20 +364,38 @@ def _join_table_by_indices(self, reduced_extra, indices):
...
+
         tpe = object if object in (left.dtype, right.dtype) else left.dtype
```
Comment (jerneju) [inline Orange/widgets/data/owmergedata.py]: No. In that case all tests fail.

File: Orange/widgets/data/owmergedata.py [hunk 4]
```diff
@@ -362,20 +364,38 @@ def _join_table_by_indices(self, reduced_extra, indices):
...
+            arr1[:, [sc for sc in string_cols if sc < left_width]] = ""
+            arr2[:, [sc - left_width for sc in string_cols if sc >= left_width]] = ""
+
+        for ind, to_change, lookup in (
+                (indices[0], arr1, left),
+                (indices[1], arr2, right)):
+            if not is_sparse:
+                known = ind != -1
+                if sum(known):
+                    to_change[known] = lookup[ind[known]]
```
Comment (kernc) [inline Orange/widgets/data/owmergedata.py]: Re lint, does this work?
```py
def _join_array_by_indices(left, right, indices, string_cols=None):

    def prepare(arr, inds):
        try:
            newarr = arr[inds]
        except IndexError:
            newarr = np.full_like(arr, np.nan)
        else:
            newarr[inds == -1] = np.full(arr.shape[1], np.nan)
        return newarr

    res = hstack((prepare(left, indices[0]),
                  prepare(right, indices[1]))
    if string_cols:
        res[:, string_cols] = ""
    return  res
```

Comment (jerneju) [inline Orange/widgets/data/owmergedata.py]: No, that does not work.

File: Orange/widgets/data/owmergedata.py [hunk 5]
```diff
@@ -362,20 +364,42 @@ def _join_table_by_indices(self, reduced_extra, indices):
...
-            arr[:, string_cols] = ""
-        for indices, to_change, lookup in (
-                (indices[0], arr[:, :left_width], left),
-                (indices[1], arr[:, left_width:], right)):
-            known = indices != -1
-            to_change[known] = lookup[indices[known]]
+            arr1[:, [sc for sc in string_cols if sc < left_width]] = ""
+            arr2[:, [sc - left_width for sc in string_cols if sc >= left_width]] = ""
+
+        for ind, to_change, lookup in (
```
Comment (kernc) [inline Orange/widgets/data/owmergedata.py]: This works!
```py
    def _join_array_by_indices(left, right, indices, string_cols=None):
        def prepare(arr, inds, str_cols):
            try:
                newarr = arr[inds]
            except IndexError:
                newarr = np.full_like(arr, np.nan)
            else:
                empty = np.full(arr.shape[1], np.nan)
                if str_cols:
                    assert arr.dtype == object
                    empty = empty.astype(object)
                    empty[str_cols] = ''
                newarr[inds == -1] = empty
            return newarr

        left_width = left.shape[1]
        str_left = [i for i in string_cols or () if i < left_width]
        str_right = [i - left_width for i in string_cols or () if i >= left_width]
        res = hstack((prepare(left, indices[0], str_left),
                      prepare(right, indices[1], str_right)))
        return res
```

**Comment by jerneju** [comment]
Well-done, @kernc !

**Comment by ajdapretnar** [comment]
@nikicc Could you please check and merge if this is done? :)

**Pull request merged**